README.md
# Fusetunnel

Fusetunnel exposes your localhost to the world for easy testing and sharing! No need to mess with DNS or deploy just to have others test out your changes.

Great for working with browser testing tools like browserling or external api callback services like twilio which require a public url for callbacks.

## Quickstart

```
npx ft --port 8000
```

## Installation

### Globally

```
npm install -g @fusebit/tunnel
```

### As a dependency in your project

```
yarn add @fusebit/tunnel
```

## CLI usage

When `Fusetunnel` is installed globally, just use the `ft` command to start the tunnel.

```
ft --port 8000
```

Thats it! It will connect to the tunnel server, setup the tunnel, and tell you what url to use for your testing. This url will remain active for the duration of your session; so feel free to share it with others for happy fun time!

You can restart your local server all you want, `ft` is smart enough to detect this and reconnect once it is back.

### Arguments

Below are some common arguments. See `ft --help` for additional arguments

- `--subdomain` request a named subdomain on the Fusetunnel server (default is random characters)
- `--local-host` proxy to a hostname other than localhost

You may also specify arguments via env variables. E.x.

```
PORT=3000 ft
```

## API

The Fusetunnel client is also usable through an API (for test integration, automation, etc)

### Fusetunnel(port [,options][,callback])

Creates a new Fusetunnel to the specified local `port`. Will return a Promise that resolves once you have been assigned a public fusetunnel url. `options` can be used to request a specific `subdomain`. A `callback` function can be passed, in which case it won't return a Promise. This exists for backwards compatibility with the old Node-style callback API. You may also pass a single options object with `port` as a property.

```js
const fusetunnel = require('tunnel');

(async () => {
  const tunnel = await fusetunnel({ port: 3000 });

  // the assigned public url for your tunnel
  // i.e. https://abcdefgjhij.tunnel.dev.fusebit.io
  tunnel.url;

  tunnel.on('close', () => {
    // tunnels are closed
  });
})();
```

#### options

- `port` (number) [required] The local port number to expose through fusetunnel.
- `subdomain` (string) Request a specific subdomain on the proxy server. **Note** You may not actually receive this name depending on availability.
- `host` (string) URL for the upstream proxy server. Defaults to `https://tunnel.dev.fusebit.io`.
- `local_host` (string) Proxy to this hostname instead of `localhost`. This will also cause the `Host` header to be re-written to this value in proxied requests.
- `local_https` (boolean) Enable tunneling to local HTTPS server.
- `local_cert` (string) Path to certificate PEM file for local HTTPS server.
- `local_key` (string) Path to certificate key file for local HTTPS server.
- `local_ca` (string) Path to certificate authority file for self-signed certificates.
- `allow_invalid_cert` (boolean) Disable certificate checks for your local HTTPS server (ignore cert/key/ca options).

Refer to [tls.createSecureContext](https://nodejs.org/api/tls.html#tls_tls_createsecurecontext_options) for details on the certificate options.

### Tunnel

The `tunnel` instance returned to your callback emits the following events

| event   | args | description                                                                          |
| ------- | ---- | ------------------------------------------------------------------------------------ |
| request | info | fires when a request is processed by the tunnel, contains _method_ and _path_ fields |
| error   | err  | fires when an error happens on the tunnel                                            |
| close   |      | fires when the tunnel has closed                                                     |

The `tunnel` instance has the following methods

| method | args | description      |
| ------ | ---- | ---------------- |
| close  |      | close the tunnel |

## server

See [fusebit/tunnel-server](https://github.com/fusebit/tunnel-server) for details on the server that powers fusetunnel.

## License

MIT


CHANGELOG.md
# 3.2.6 (2021-6-3)
- Fusetunnel remote connection retry timeout.
# 3.2.5 (2021-05-25)
- Fusetunnel now handles close properly. 

# 3.2.4 (2021-05-24)
- Add the clientId to the Tunnel type definition.

# 3.2.3 (2021-05-20)
- Switch prettier to dev dependency.

# 3.2.2 (2021-05-20)
- Switch to a manual calculation for sha256 fingerprint to ensure backward compatibility.

# 3.2.1 (2021-02-22)
- Removed pings to server as they violate correct use of the tcp socket passed to httpAgent

# 3.2.0 (2021-02-20)
- Added TS typings
- Rebranded to "mytunnel"
- Added pings to server to help with dead connections detection.

# 2.0.1 (2021-01-09)

- Upgrade dependencies

# 2.0.0 (2019-09-16)

- Add support for tunneling a local HTTPS server
- Add support for localtunnel server with IP-based tunnel URLs
- Node.js client API is now Promise-based, with backwards compatibility to callback
- Major refactor of entire codebase using modern ES syntax (requires Node.js v8.3.0 or above)

# 1.9.2 (2019-06-01)

- Update debug to 4.1.1
- Update axios to 0.19.0

# 1.9.1 (2018-09-08)

- Update debug to 2.6.9

# 1.9.0 (2018-04-03)

- Add _request_ event to Tunnel emitter
- Update yargs to support config via environment variables
- Add basic request logging when --print-requests argument is used

# 1.8.3 (2017-06-11)

- update request dependency
- update debug dependency
- update openurl dependency

# 1.8.2 (2016-11-17)

- fix host header transform
- update request dependency

# 1.8.1 (2016-01-20)

- fix bug w/ HostHeaderTransformer and binary data

# 1.8.0 (2015-11-04)

- pass socket errors up to top level

# 1.7.0 (2015-07-22)

- add short arg options

# 1.6.0 (2015-05-15)

- keep sockets alive after connecting
- add --open param to CLI

# 1.5.0 (2014-10-25)

- capture all errors on remote socket and restart the tunnel

# 1.4.0 (2014-08-31)

- don't emit errors for ETIMEDOUT

# 1.2.0 / 2014-04-28

- return `client` from `localtunnel` API instantiation

# 1.1.0 / 2014-02-24

- add a host header transform to change the 'Host' header in requests

# 1.0.0 / 2014-02-14

- default to localltunnel.me for host
- remove exported `connect` method (just export one function that does the same thing)
- change localtunnel signature to (port, opt, fn)

# 0.2.2 / 2014-01-09


fusetunnel.spec.js
/* eslint-disable no-console */

const crypto = require('crypto');
const http = require('http');
const https = require('https');
const url = require('url');
const supertest = require('supertest');
const chai = require('chai');
chai.use(require('chai-string'));
const expect = chai.expect;
const assert = require('assert');

const localtunnel = require('./localtunnel');

let fakePort;
let testServer;

describe('localtunnel', () => {
  before((done) => {
    testServer = http.createServer();
    testServer.on('request', (req, res) => {
      res.write(req.headers.host);
      res.end();
    });
    testServer.listen(() => {
      const { port } = testServer.address();
      fakePort = port;
      done();
    });
  });

  after(() => {
    testServer.close();
  });

  it('query localtunnel server w/ ident', async () => {
    const tunnel = await localtunnel({ port: fakePort });

    try {
      await supertest(tunnel.url)
        .get('/')
        .expect(200)
        .expect((res) => {
          expect(tunnel.url).to.endsWith(res.text);
        });
    } finally {
      tunnel.close();
    }
  });

  it('request specific domain', async () => {
    const subdomain = Math.random().toString(36).substr(2);
    const tunnel = await localtunnel({ port: fakePort, subdomain });
    tunnel.close();

    expect(tunnel.url).to.startsWith(`https://${subdomain}.`);
  });

  describe('--local-host localhost', () => {
    it('override Host header with local-host', async () => {
      const tunnel = await localtunnel({ port: fakePort, local_host: 'localhost' });

      try {
        await supertest(tunnel.url)
          .get('/')
          .expect(200)
          .expect((res) => {
            expect(res.text).to.equal('localhost');
          });
      } finally {
        tunnel.close();
      }
    });
  });

  describe('--local-host 127.0.0.1', () => {
    it('override Host header with local-host', async () => {
      const tunnel = await localtunnel({ port: fakePort, local_host: '127.0.0.1' });

      try {
        await supertest(tunnel.url)
          .get('/')
          .expect(200)
          .expect((res) => {
            expect(res.text).to.equal('127.0.0.1');
          });
      } finally {
        tunnel.close();
      }
    });

    it('send chunked request', async () => {
      const tunnel = await localtunnel({ port: fakePort, local_host: '127.0.0.1' });

      const parsed = url.parse(tunnel.url);
      const opt = {
        host: parsed.host,
        port: 443,
        headers: {
          host: parsed.hostname,
          'Transfer-Encoding': 'chunked',
        },
        path: '/',
      };

      await new Promise((resolve, reject) => {
        const req = https.request(opt, (res) => {
          res.setEncoding('utf8');
          let body = '';

          res.on('data', (chunk) => {
            body += chunk;
          });

          res.on('end', () => {
            try {
              assert.strictEqual(body, '127.0.0.1');
            } catch (e) {
              reject(e);
            } finally {
              tunnel.close();
            }
            resolve();
          });
        });

        req.end(crypto.randomBytes(1024 * 8).toString('base64'));
      });
    });
  });
});


lib/Tunnel.js
/* eslint-disable consistent-return, no-underscore-dangle */

const { parse } = require('url');
const { EventEmitter } = require('events');
const axios = require('axios');
const debug = require('debug')('mytunnel:client');

const TunnelCluster = require('./TunnelCluster');

module.exports = class Tunnel extends EventEmitter {
  constructor(opts = {}) {
    super(opts);
    this.opts = opts;
    this.closed = false;
    if (!this.opts.host) {
      this.opts.host = 'https://tunnel.dev.fusebit.io';
    }
  }

  _getInfo(body) {
    /* eslint-disable camelcase */
    const { id, ip, port, url, cached_url, max_conn_count, isTls, certificate, OverrideTunnelIp } = body;
    const { host, port: local_port, local_host } = this.opts;
    const { local_https, local_cert, local_key, local_ca, allow_invalid_cert } = this.opts;
    return {
      name: id,
      url,
      cached_url,
      max_conn: max_conn_count || 1,
      remote_host: parse(host).hostname,
      remote_ip: ip,
      remote_port: port,
      local_port,
      local_host,
      local_https,
      local_cert,
      local_key,
      local_ca,
      allow_invalid_cert,
      isTls,
      certificate,
      OverrideTunnelIp,
    };
    /* eslint-enable camelcase */
  }

  // initialize connection
  // callback with connection info
  _init(cb) {
    const opt = this.opts;
    const getInfo = this._getInfo.bind(this);

    const params = {
      responseType: 'json',
    };

    const baseUri = `${opt.host}/`;
    // no subdomain at first, maybe use requested domain
    const assignedDomain = opt.subdomain;
    // where to quest
    const uri = baseUri + (assignedDomain || '?new');
    let totalRetries = 0;
    (function getUrl() {
      axios
        .get(uri, params)
        .then((res) => {
          const body = res.data;
          debug('got tunnel information', res.data);
          if (res.status !== 200) {
            const err = new Error((body && body.message) || 'localtunnel server returned an error, please try again');
            return cb(err);
          }
          cb(null, getInfo(body));
        })
        .catch((err) => {
          totalRetries++;
          debug(`tunnel server offline: ${err.message}, retry 1s`);
          if (totalRetries >= 5) {
            debug(
              `Tunnel server api tried to reconnect more than 5 times, something probably failed miserably, terminating the tunnel.`
            );
            throw Error('Terminate Tunnel');
          }
          return setTimeout(getUrl, 1000);
        });
    })();
  }

  _establish(info) {
    // increase max event listeners so that localtunnel consumers don't get
    // warning messages as soon as they setup even one listener. See #71
    this.setMaxListeners(info.max_conn + (EventEmitter.defaultMaxListeners || 10));

    this.tunnelCluster = new TunnelCluster(info);

    // only emit the url the first time
    this.tunnelCluster.once('open', () => {
      this.emit('url', info.url);
    });

    // re-emit socket error
    this.tunnelCluster.on('error', (err) => {
      debug('got socket error', err.message);
      this.emit('error', err);
    });

    let tunnelCount = 0;

    // track open count
    this.tunnelCluster.on('open', (tunnel) => {
      tunnelCount++;
      debug('tunnel open [total: %d]', tunnelCount);

      const closeHandler = () => {
        tunnel.destroy();
      };

      if (this.closed) {
        return closeHandler();
      }

      this.once('close', closeHandler);
      tunnel.once('close', () => {
        this.removeListener('close', closeHandler);
      });
    });

    // when a tunnel dies, open a new one
    this.tunnelCluster.on('dead', () => {
      tunnelCount--;
      debug('tunnel dead [total: %d]', tunnelCount);
      if (this.closed) {
        return;
      }
      this.tunnelCluster.open();
    });

    this.tunnelCluster.on('request', (req) => {
      this.emit('request', req);
    });

    // establish as many tunnels as allowed
    for (let count = 0; count < info.max_conn; ++count) {
      this.tunnelCluster.open();
    }
  }

  open(cb) {
    this._init((err, info) => {
      if (err) {
        return cb(err);
      }

      this.clientId = info.name;
      this.url = info.url;

      // `cached_url` is only returned by proxy servers that support resource caching.
      if (info.cached_url) {
        this.cachedUrl = info.cached_url;
      }

      this._establish(info);
      cb();
    });
  }

  close() {
    this.closed = true;
    this.emit('close');
  }
};


lib/TunnelCluster.js
const { EventEmitter } = require('events');
const debug = require('debug')('mytunnel:client');
const fs = require('fs');
const net = require('net');
const tls = require('tls');
const constants = require('constants');
const HeaderHostTransformer = require('./HeaderHostTransformer');
const crypto = require('crypto');
tls.DEFAULT_ECDH_CURVE = 'auto';
// manages groups of tunnels
module.exports = class TunnelCluster extends EventEmitter {
  constructor(opts = {}) {
    super(opts);
    this.opts = opts;
  }

  open() {
    const opt = this.opts;

    // Prefer IP if returned by the server
    const remoteHostOrIp = opt.OverrideTunnelIp || opt.remote_ip || opt.remote_host;
    const remotePort = opt.remote_port;
    const localHost = opt.local_host || 'localhost';
    const localPort = opt.local_port;
    const localProtocol = opt.local_https ? 'https' : 'http';
    const allowInvalidCert = opt.allow_invalid_cert;
    const isTls = opt.isTls;
    const cert = opt.certificate;
    debug('establishing tunnel %s://%s:%s <> %s:%s', localProtocol, localHost, localPort, remoteHostOrIp, remotePort);
    let remote;
    if (isTls) {
      remote = tls.connect({
        host: remoteHostOrIp,
        port: remotePort,
        rejectUnauthorized: false,
        minVersion: 'TLSv1.3',
      });
    } else {
      remote = net.connect({
        host: remoteHostOrIp,
        port: remotePort,
      });
    }

    remote.setKeepAlive(true);
    remote.on('error', (err) => {
      debug('got remote connection error', err.message);
      // emit connection refused errors immediately, because they
      // indicate that the tunnel can't be established.
      if (err.code === 'ECONNREFUSED') {
        this.emit(
          'error',
          new Error(`connection refused: ${remoteHostOrIp}:${remotePort} (check your firewall settings)`)
        );
      }

      remote.end();
    });

    const connLocal = () => {
      if (remote.destroyed) {
        debug('remote destroyed');
        this.emit('dead');
        return;
      }

      debug('connecting locally to %s://%s:%d', localProtocol, localHost, localPort);
      remote.pause();

      if (allowInvalidCert) {
        debug('allowing invalid certificates');
      }

      const getLocalCertOpts = () =>
        allowInvalidCert
          ? { rejectUnauthorized: false }
          : {
              cert: fs.readFileSync(opt.local_cert),
              key: fs.readFileSync(opt.local_key),
              ca: opt.local_ca ? [fs.readFileSync(opt.local_ca)] : undefined,
            };

      // connection to local http server
      const local = opt.local_https
        ? tls.connect({
            host: localHost,
            port: localPort,
            ...getLocalCertOpts(),
          })
        : net.connect({ host: localHost, port: localPort });

      const remoteClose = () => {
        debug('remote close');
        this.emit('dead');
        local.end();
      };

      remote.once('close', remoteClose);

      // TODO some languages have single threaded servers which makes opening up
      // multiple local connections impossible. We need a smarter way to scale
      // and adjust for such instances to avoid beating on the door of the server
      local.once('error', (err) => {
        debug('local error %s', err.message);
        local.end();

        remote.removeListener('close', remoteClose);

        if (err.code !== 'ECONNREFUSED') {
          return remote.end();
        }

        // retrying connection to local server
        setTimeout(connLocal, 1000);
      });

      local.once('connect', () => {
        debug('connected locally');
        remote.resume();

        let stream = remote;

        // if user requested specific local host
        // then we use host header transform to replace the host header
        if (opt.local_host) {
          debug('transform Host header to %s', opt.local_host);
          stream = remote.pipe(new HeaderHostTransformer({ host: opt.local_host }));
        }

        stream.pipe(local).pipe(remote);

        // when local closes, also get a new remote
        local.once('close', (hadError) => {
          debug('local connection closed [%s]', hadError);
        });
      });
    };

    remote.on('data', (data) => {
      const match = data.toString().match(/^(\w+) (\S+)/);
      if (match) {
        this.emit('request', {
          method: match[1],
          path: match[2],
        });
      }
    });

    // tunnel is considered open when remote connects
    remote.once(isTls ? 'secureConnect' : 'connect', () => {
      if (isTls) {
        debug('TLS is enabled');
        const baseString = cert.match(/-----BEGIN CERTIFICATE-----\s*([\s\S]+?)\s*-----END CERTIFICATE-----/i);
        const rawCert = Buffer.from(baseString[1], 'base64');
        const sha256sum = crypto.createHash('sha256').update(rawCert).digest('hex');
        const fingerprint = sha256sum.toUpperCase().replace(/(.{2})(?!$)/g, '$1:');
        // eg 83:6E:3E:99:58:44:AE:61:72:55:AD:C6:24:BE:5C:2D:46:21:BA:BE:87:E4:3A:38:C8:E8:09:AC:22:48:46:20
        try {
          if (fingerprint !== remote.getPeerCertificate().fingerprint256) {
            this.emit('error', {
              message: 'tls certificate does not match what was sent over',
            });
            return;
          }
        } catch {
          this.emit('error', {
            message: 'tls socket closed too fast, was not able to verify the fingerprint',
          });
          return;
        }
      }
      this.emit('open', remote);
      connLocal();
    });
  }
};
